File: schemas/social-media/drizzle/reposts.ts
// reposts: Shares of existing posts to a user's timeline.
// See README.md for full design rationale.
import { pgTable, uuid, timestamp, index, unique } from "drizzle-orm/pg-core";
import { posts } from "./posts";
import { users } from "./users";

export const reposts = pgTable(
  "reposts",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    postId: uuid("post_id")
      .notNull()
      .references(() => posts.id, { onDelete: "cascade" }),
    userId: uuid("user_id")
      .notNull()
      .references(() => users.id, { onDelete: "cascade" }),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    unique("uq_reposts_post_user").on(table.postId, table.userId),
    index("idx_reposts_user_id_created_at").on(table.userId, table.createdAt),
  ]
);

File: schemas/social-media/spacetimedb/reposts.rs
// reposts: user reposts (shares) of existing posts.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[spacetimedb::table(name = reposts, public)]
pub struct Repost {
    #[primary_key]
    pub id: String, // UUID
    // Composite unique: (post_id, user_id) — enforce in reducer logic.
    #[index(btree)]
    pub post_id: String, // UUID — FK → posts.id (cascade delete)
    // Composite index: (user_id, created_at) — not supported in SpacetimeDB; document only.
    #[index(btree)]
    pub user_id: String, // UUID — FK → users.id (cascade delete)
    pub created_at: Timestamp,
}
